core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -643,7 +645,12 @@ private[spark] object Utils extends Logging {
           fileOverwrite)
       } else {
         val in = fs.open(innerPath)
-        val targetFile = new File(targetDir, innerPath.getName)
+        val targetFile: File =
+          if (!filename.isEmpty) {
+            new File(targetDir, filename)
+          } else {
+            new File(targetDir, innerPath.getName)
+          }
```

Comment:
> If the path passed to it is a directory, it will recursively call itself without setting filename

That's not actually true. If you call `listStatus` on a directory it will list the directory's contents and not include the directory itself. So if the directory contains multiple files they will all go into the else case in L646 and be renamed to the same thing.